### ISSUE TITLE
Initializing @OneToMany with mutableListOf() in Kotlin results in ClassCastException

### ISSUE THREAD

**Issue by davidhiendl** [ebean-orm/ebean#2404]
When initializing fields like the following a ClassCastException is thrown:
```kotlin
    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL])
    var items: MutableList<OrderItem> = mutableListOf()
```

It is undesirable to have fields set to be nullable like the following in my opinion, because it can make code that uses the models much more verbose:
```kotlin
    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL])
    var items: MutableList<OrderItem>? = null
```

Am I missing something important here or is the intended behavior to use nullable relationship fields?

## Expected behavior
Initializing fields with default values should not result in ClassCastException.

## Actual behavior
When fields are initialized similar to the example code below a ClassCastException will be thrown.

### Steps to reproduce
This example code is an excerpt of actual application code that triggers the issue.

```kotlin
@Entity
@Table(name = "sales_order")
class Order {

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL])
    var items: MutableList<OrderItem> = mutableListOf()

}

@Entity
@Table(name = "sales_order_item")
class OrderItem {

    @ManyToOne(optional = false)
    @DbForeignKey(onDelete = ConstraintMode.CASCADE)
    var order: Order? = null

}

class OrderController {

    fun read() {
        val order = QOrder()
            .id.eq(123)
            .apply {
                items.fetch() // this line triggers the error, it does not occur when not eagerly fetching data!
            }
            .findOne() // error stack trace starts here
    }

}


```

```console
16:55:04.653 [io-executor-thread-1] ERROR io.micronaut.http.server.netty.RoutingInBoundHandler - Unexpected error occurred: class java.util.ArrayList cannot be cast to class io.ebean.bean.BeanCollection (java.util.ArrayList is in module java.base of loader 'bootstrap'; io.ebean.bean.BeanCollection is in unnamed module of loader 'app')
java.lang.ClassCastException: class java.util.ArrayList cannot be cast to class io.ebean.bean.BeanCollection (java.util.ArrayList is in module java.base of loader 'bootstrap'; io.ebean.bean.BeanCollection is in unnamed module of loader 'app')
	at io.ebeaninternal.server.deploy.BeanPropertyAssocMany.addBeanToCollectionWithCreate(BeanPropertyAssocMany.java:229) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.SqlTreeNodeManyRoot.load(SqlTreeNodeManyRoot.java:35) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.SqlTreeNodeBean$Load.loadChildren(SqlTreeNodeBean.java:312) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.SqlTreeNodeBean$Load.initialise(SqlTreeNodeBean.java:416) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.SqlTreeNodeBean$Load.perform(SqlTreeNodeBean.java:423) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.SqlTreeNodeBean.load(SqlTreeNodeBean.java:446) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.SqlTreeNodeRoot.load(SqlTreeNodeRoot.java:43) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.CQuery.readNextBean(CQuery.java:415) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.CQuery.hasNext(CQuery.java:490) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.CQuery.readBean(CQuery.java:464) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.CQueryEngine.find(CQueryEngine.java:391) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.query.DefaultOrmQueryEngine.findId(DefaultOrmQueryEngine.java:175) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.core.OrmQueryRequest.findId(OrmQueryRequest.java:364) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.core.DefaultServer.findId(DefaultServer.java:1155) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.core.DefaultServer.findOne(DefaultServer.java:1172) ~[ebean-core-12.12.0.jar:?]
	at io.ebeaninternal.server.querydefn.DefaultOrmQuery.findOne(DefaultOrmQuery.java:1508) ~[ebean-core-12.12.0.jar:?]
	at io.ebean.typequery.TQRootBean.findOne(TQRootBean.java:1618) ~[ebean-querybean-12.12.0.jar:?]
	at de.dhswt.bsmxcd.webapi.sales.order.OrderController.read(OrderController.kt:72) ~[webapi.jar:?]
	at de.dhswt.bsmxcd.webapi.sales.order.$OrderControllerDefinition$$exec2.invokeInternal(Unknown Source) ~[webapi.jar:?]
	at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:151) ~[micronaut-inject-2.5.13.jar:2.5.13]
	at io.micronaut.context.DefaultBeanContext$4.invoke(DefaultBeanContext.java:482) ~[micronaut-inject-2.5.13.jar:2.5.13]
	at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:303) ~[micronaut-router-2.5.13.jar:2.5.13]
	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:121) ~[micronaut-router-2.5.13.jar:2.5.13]
	at io.micronaut.http.server.netty.RoutingInBoundHandler.emitRouteResponse(RoutingInBoundHandler.java:1502) ~[micronaut-http-server-netty-2.5.13.jar:2.5.13]
	at io.micronaut.http.server.netty.RoutingInBoundHandler.access$2100(RoutingInBoundHandler.java:151) ~[micronaut-http-server-netty-2.5.13.jar:2.5.13]
	at io.micronaut.http.server.netty.RoutingInBoundHandler$6$1.request(RoutingInBoundHandler.java:1469) ~[micronaut-http-server-netty-2.5.13.jar:2.5.13]
	at io.reactivex.internal.subscriptions.SubscriptionHelper.deferredSetOnce(SubscriptionHelper.java:202) ~[rxjava-2.2.10.jar:?]
	at io.reactivex.internal.operators.single.SingleFlatMapPublisher$SingleFlatMapPublisherObserver.onSubscribe(SingleFlatMapPublisher.java:107) ~[rxjava-2.2.10.jar:?]
	at io.micronaut.http.server.netty.RoutingInBoundHandler$6.doSubscribe(RoutingInBoundHandler.java:1457) ~[micronaut-http-server-netty-2.5.13.jar:2.5.13]
	at io.micronaut.http.server.netty.RoutingInBoundHandler$6.lambda$subscribe$0(RoutingInBoundHandler.java:1451) ~[micronaut-http-server-netty-2.5.13.jar:2.5.13]
	at io.micronaut.scheduling.instrument.InvocationInstrumenterWrappedRunnable.run(InvocationInstrumenterWrappedRunnable.java:47) [micronaut-context-2.5.13.jar:2.5.13]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]

```

**Comment by rbygrave** [comment]
The bytecode generated by Kotlin for initialising a collection has changed (looks like a kotlin bug to be honest).

The bytecode in ASM form looks like:

```java
  mv.visitVarInsn(ALOAD, 0);
  mv.visitInsn(ICONST_0);        // THIS IS DIFFERENT, not expected
  mv.visitVarInsn(ISTORE, 1);   // THIS IS DIFFERENT, not expected
  mv.visitTypeInsn(NEW, "java/util/ArrayList");
  mv.visitInsn(DUP);
  mv.visitMethodInsn(INVOKESPECIAL, "java/util/ArrayList", "<init>", "()V", false);
  mv.visitTypeInsn(CHECKCAST, "java/util/List");  // Optional: only in Kotlin generated bytecode, not in Java bytecode
  mv.visitFieldInsn(PUTFIELD, "org/foo/Hello", "lines", "Ljava/util/List;");
```

We can see the Kotlin bytecode now has the additional instructions (which are unnecessary and unexpected).
```java
  mv.visitInsn(ICONST_0);       
  mv.visitVarInsn(ISTORE, 1);  
```
Regardless, we take into the additional CHECKCAST that kotlin also emits (relative to Java bytecode).

The fix for this is to take into account that Kotlin can now generate these additional bytecodes.  

The background here is that with Ebean we need to ensure that `@OneToMany` and `@ManyToMany` collections are initialised by ebean in order to properly support orphan removal etc.  As such the ebean-agent detects the initialisation bytecode for these collections and effectively removes that bytecode thus leaving the initialisation of these collections to ebean (which will initialise them as BeanCollection implementations that are aware of removals and additions etc).

With this change to the bytecode generated by Kotlin we need to adjust the ebean-agent to also account for the additional bytecode instructions when detecting initialisation of entity collections (`@OneToMany` and `@ManyToMany` collections).

**Comment by davidhiendl** [comment]
I see, I didn't realize this was tied to the Kotlin version as I only noticed it in a new Project

For now and anyone looking for a temporary workaround: `lateint var` with private set has identical semantics to val or without private set to var.

**Comment by rbygrave** [comment]
> tied to the Kotlin version as I only noticed it in a new project.

Yes, the bytecode generated by kotlin has changed just a little bit. It produces extra bytecode that javac does not - seems a bit unncessary but regardless I have modified the ebean-agent to deal with that now.  

I have released version **12.12.1** of ebean-agent, ebean-maven-plugin, io.ebean.tile:enhancement and the ebean-gradle-plugin. Bumping to that will fix this issue.
